packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +460,347 @@ class ShellRoute extends RouteBase {
     }
   }
 
-  /// The widget builder for a shell route.
+  /// The [GlobalKey] to be used by the [Navigator] built for this route.
+  /// All ShellRoutes build a Navigator by default. Child GoRoutes
+  /// are placed onto this Navigator instead of the root Navigator.
+  final GlobalKey<NavigatorState> navigatorKey;
+
+  /// Restoration ID to save and restore the state of the navigator, including
+  /// its history.
+  final String? restorationScopeId;
+}
+
+/// A route that displays a UI shell with separate [Navigator]s for its
+/// sub-routes.
+///
+/// Similar to [ShellRoute], this route class places its sub-route on a
+/// different Navigator than the root Navigator. However, this route class
+/// differs in that it creates separate Navigators for each of its nested
+/// branches (i.e. parallel navigation trees), making it possible to build an
+/// app with stateful nested navigation. This is convenient when for instance
+/// implementing a UI with a [BottomNavigationBar], with a persistent navigation
+/// state for each tab.
+///
+/// A StatefulShellRoute is created by providing a List of [StatefulShellBranch]
+/// items, each representing a separate stateful branch in the route tree. The
+/// branches can be provided either statically, by passing a list of branches in
+/// the constructor, or dynamically by instead providing a [branchBuilder].
```

Comment:
maybe dynamic generation of branches without having to create a new `GoRouter` ?

Reply:
This is interesting, I was actually getting a similar request that people want to dynamically adding new sub-routes to a goRoute object. we maybe need to come up with something more systematic

Reply:
Yes, the reason for branchBuilder is to support a dynamic set of branches. But an alternative could of course be to add the possibility to modify the existing route configuration.

Reply:
I think we should just provide a way to update RouteBase table during runtime. and remove this API WDYT?

Reply:
If we can find a good and convenient way of replacing/modifying the routes of GoRouter/RouteConfiguration, then I believe that might be even better. I'll do some experimenting with this concept to see how that works out. 

By the way (in case I've failed to mention it before), there is also an additional stateful shell route example that showcases the use of dynamic branches: `stateful_shell_route_dynamic.dart`.

Reply:
It can be a separate pr though since this pr is already huge

Reply:
Yes, I think so too, I only meant I wanted to do a quick test to see how such a change could provide a good solution for dynamic branches. 

But I think I'll primarily focus on removing/reverting the changes for dynamic routes for now, as well as fixing the other review issues, so we can hopefully come to a close with this PR soon. And I'm all for streamlining (less is more after all) and postponing some features that are not at the core of the solution (hopefully most of the other features beside this are though).